src/experiment.ts

```diff
@@ -0,0 +1,125 @@
+import { Identify } from 'amplitude-js';
+import { Client } from './client';
+
+const LOCAL_STORAGE_EXPERIMENTS_PREFIX = '__analytics_experiments_';
+
+/** Experiment interface allows getting a variation for a specified user. */
+export interface Experiment<Variation extends string> {
+	name: string;
+	engage(userId: string): Variation;
+}
+
+interface VariationData<Variation> {
+	variation: Variation;
+	targetPercent: number;
+}
+
+/** LocalExperiment allows describing an Experiment implemented with local storage. */
+export class LocalExperiment<Variation extends string>
+	implements Experiment<Variation> {
+	private readonly data: Array<VariationData<Variation>> = [];
+	private coveredPercent: number = 0;
+
+	constructor(
+		public readonly name: string,
+		private readonly analytics?: Client,
+	) {}
+
+	private static checkPercent(f: number) {
+		if (isNaN(f) || f < 0 || f > 100) {
+			throw new Error(
+				'Variation fraction must be defines as a percent value between 0 and 100',
+			);
+		}
+	}
+
+	private checkDuplicates(variation: Variation) {
+		const present = this.data.find(d => d.variation === variation);
+		if (present != null) {
+			throw new Error(
+				`Variation [${present.variation} ${present.targetPercent}%] already exists in experiment ${this.name}.`,
+			);
+		}
+	}
+
+	private static dataString(data: Array<VariationData<any>>): string {
+		return data
+			.map(d => `variation ${d.variation}: ${d.targetPercent}%`)
+			.join(', ');
+	}
+
+	define(
+		variation: Variation,
+		targetPercent: number,
+	): LocalExperiment<Variation> {
+		LocalExperiment.checkPercent(targetPercent);
```

Comment:
Let's just take `checkPercent` outside the class

Reply:
Done.